### PR TITLE
22457-Provide-a-CalendarPresenter-fully-written-in-Spec

### PR DESCRIPTION
**Pull request by juliendelplanque**
Implemented CalendarPresenter and its unit tests.

Here is how it looks like (dark theme):

calendar

### CODE REVIEW

**Comment by juliendelplanque** [comment]
Could someone restart the CI? I think the failure on unix is not related to this change since no other platform complains?

**Comment by pavel-krivanek** [comment]
The date of the week does not work well. For 17.9.2018 it shows Wed. Some customization of the calendar that respects the locale would be nice.

**Comment by juliendelplanque** [comment]
Do you have examples of such customisations?

**Comment by juliendelplanque** [comment]
@pavel-krivanek, I fixed the bug and added additional unit tests? Could you re-review this PR and integrate it?

**Review by pavel-krivanek — APPROVED** [review]

File: src/Spec-PolyWidgets/CalendarPresenter.class.st [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+{ #category : #private }
+CalendarPresenter >> datesToDisplayFor: aDate [
+	"Returns the dates to display in the calendar."
+	^ (self julianDaysIntervalFor: aDate)
+			collect: [ :julianNumber | Date julianDayNumber: julianNumber ]
+]
+
+{ #category : #private }
+CalendarPresenter >> dayBeforeMonthOf: aDate [
+	^ aDate month dates first dayOfWeek - 1
```
Comment (dionisiydk) [inline src/Spec-PolyWidgets/CalendarPresenter.class.st:85]: Such kind of logic should be moved to date itself. It is generally usefull and should not be hidden inside UI

File: src/Spec-PolyWidgets/CalendarPresenter.class.st [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+]
+
+{ #category : #'private - event' }
+CalendarPresenter >> dayClicked: aDate [
+	self whenDaySelectedBlock cull: aDate cull: self
+]
+
+{ #category : #'private - constants' }
+CalendarPresenter >> dayNames [
+	^ #('Sun' 'Mon' 'Tue' 'Wed' 'Thu' 'Fri' 'Sat')
```
Comment (dionisiydk) [inline src/Spec-PolyWidgets/CalendarPresenter.class.st:95]: There are Week dayNames or similar.

File: src/Spec-PolyWidgets/CalendarPresenter.class.st [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+{ #category : #private }
+CalendarPresenter >> daysButtonsBuilderFrom: aDate [
+	^ DynamicPresentersListBuilder new
+		modelObjects: (self datesToDisplayFor: aDate);
+		presenter: ButtonPresenter
+		configuredAs: [ :button :date | 
+			button
+				label: date dayOfMonth asString;
+				enabled: aDate monthIndex = date monthIndex;
```
Comment (dionisiydk) [inline src/Spec-PolyWidgets/CalendarPresenter.class.st:121]: better would be 
```Smalltalk
aDate isAtSameMonthWith: date
```

File: src/Spec-PolyWidgets/CalendarPresenter.class.st [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+				action: [ self dayClicked: date ].
+			"Add icon next to current date."
+			date julianDayNumber = Date today julianDayNumber
```
Comment (dionisiydk) [inline src/Spec-PolyWidgets/CalendarPresenter.class.st:124]:
```Smalltalk
date isAtSameDayWith: Date today
```
or better
```Smalltalk
date isToday
```
But it needs different name to not compare month and year

File: src/Spec-PolyWidgets/CalendarPresenter.class.st [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+{ #category : #private }
+CalendarPresenter >> julianDaysIntervalFor: aDate [
+	"Returns an interval of integers with the first one being the first day to display in the UI and the last integer the last day.
+	 The first and last days might not be in the same month as aDate provided as parameter."
+
+	| start end |
+	start := aDate julianDayNumber - aDate dayOfMonth - (self dayBeforeMonthOf: aDate) + 1.
+	end := aDate julianDayNumber + (aDate month daysInMonth - aDate dayOfMonth) + (self daysAfterMonthOf: aDate).
+	^ start to: end
```
Comment (dionisiydk) [inline src/Spec-PolyWidgets/CalendarPresenter.class.st:183]: Also looks like model responsibility

**Review by dionisiydk — COMMENTED** [review]
There are many useful features which can be part of model (date related classes) and can be reused outside of UI

**Pull request merged**